zephyr/lib/vpages.c

```diff
@@ -0,0 +1,340 @@
+		return ret;
+	}
+
+	/* success update the free pages */
+	page_context.free_pages -= pages;
+
+	/* store the size and virtual page number in first free alloc element,
+	 * we have already checked for a free element before the mapping.
+	 */
+	for (int i = 0; i < VPAGE_MAX_ALLOCS; i++) {
+		if (page_context.velems[i].pages == 0) {
+			page_context.velems[i].pages = pages;
+			page_context.velems[i].vpage =
+				(POINTER_TO_UINT(vaddr) -
+				POINTER_TO_UINT(page_context.vpage_blocks.buffer)) /
+				CONFIG_MM_DRV_PAGE_SIZE;
+			page_context.num_elems_in_use++;
+			break;
+		}
+	}
+
+	/* return the virtual address */
+	*ptr = vaddr;
+	return ret;
+}
+
+/**
+ * @brief Allocate virtual memory pages
+ *
+ * Allocates virtual memory pages from the virtual page allocator.
+ *
+ * @param pages Number of 4kB pages to allocate.
+ * @retval NULL on allocation failure.
+ */
+void *alloc_vpages(uint32_t pages)
+{
+	void *ptr;
+	int err;
+
+	k_mutex_lock(&page_context.lock, K_FOREVER);
+	err = vpages_alloc_and_map(pages, &ptr);
+	k_mutex_unlock(&page_context.lock);
+	if (err < 0) {
+		LOG_ERR("vpage_alloc failed %d for %d pages, total %d free %d",
+			err, pages, page_context.total_pages, page_context.free_pages);
+	}
+	LOG_INF("vpage_alloc ptr %p pages %d free %d/%d", ptr, pages, page_context.free_pages,
+		page_context.total_pages);
+	return ptr;
+}
+
+/**
+ * @brief Free and unmap virtual memory pages
+ *
+ * Frees previously allocated virtual memory pages and unmaps them.
+ *
+ * @param ptr Pointer to the memory pages to free.
+ * @retval 0 if successful.
+ * @retval -EINVAL if ptr is invalid.
+ */
+static int vpages_free_and_unmap(uintptr_t *ptr)
+{
+	int pages = 0;
+	int ret;
+
+	/* check for valid ptr which must be page aligned */
+	if (IS_ALIGNED(ptr, CONFIG_MM_DRV_PAGE_SIZE) == 0) {
+		LOG_ERR("error: invalid non aligned page pointer %p", ptr);
+		return -EINVAL;
+	}
+
+	/* find the allocation element */
+	for (int i = 0; i < VPAGE_MAX_ALLOCS; i++) {
+		if (page_context.velems[i].pages > 0 &&
+		    page_context.velems[i].vpage ==
+		    (POINTER_TO_UINT(ptr) - POINTER_TO_UINT(page_context.vpage_blocks.buffer)) /
+		    CONFIG_MM_DRV_PAGE_SIZE) {
+
+			pages = page_context.velems[i].pages;
+
+			LOG_DBG("found allocation element %d pages %d vpage %d for ptr %p",
+				i, page_context.velems[i].pages,
+				page_context.velems[i].vpage, ptr);
+
+			/* clear the element */
+			page_context.velems[i].pages = 0;
+			page_context.velems[i].vpage = 0;
+			page_context.num_elems_in_use--;
+			break;
+		}
+	}
+
+	/* check we found allocation element */
+	if (pages == 0) {
+		LOG_ERR("error: invalid page pointer %p not found", ptr);
+		return -EINVAL;
+	}
+
+	/* unmap the pages from virtual region */
+	ret = sys_mm_drv_unmap_region((void *)ptr, pages * CONFIG_MM_DRV_PAGE_SIZE);
+	if (ret < 0) {
+		LOG_ERR("error: failed to unmap virtual region %p pages %d, error %d",
+			ptr, pages, ret);
+		return ret;
+	}
+
+	/* free physical blocks */
```

Comment:
I think the previous call frees physical pages in `sys_mm_drv_unmap_page_wflush()` by calling `sys_mem_blocks_free_contiguous(&L2_PHYS_SRAM_REGION, ...)`, this call frees the virtual address range

Reply:
I meant that the comment might be not quite correct